harp/reader.py

```diff
@@ -51,13 +51,21 @@ class DeviceReader:
     def __init__(self, device: Model, registers: dict[str, RegisterReader]) -> None:
         self.device = device
         self.registers = registers
+        self._register_map = {self.registers[key].register.address: key for key in self.registers.keys()}
```

Comment:
I agree the simplest approach seems to be creating an indirect table of values. However, to expose it through a public attribute `registers` we will probably have to create a new class here (e.g. `RegisterMap` instead of a simple `dict`). We probably should make sure to implement `__iter__` and maybe even `__len__`, `__repr__` etc to make it feel like a regular dictionary object but accepting multiple keys for each value.

I do see the convenience in making the access through `__getitem__`. The reason that keeps me from it is that this object also exposes the public `device` attribute which may be important for metadata and it might be lost in the noise if we start implementing the entire collection at the root object.